src/windows/disk.rs

```diff
@@ -19,12 +23,110 @@ use windows::Win32::System::Ioctl::{
+    buf[..=end].to_vec()
+}
+
+// Realistically, volume names are probably not longer than 44 characters,
+// but the example in the Microsoft documentation uses MAX_PATH as well.
+// https://learn.microsoft.com/en-us/windows/win32/fileio/displaying-volume-paths
+const VOLUME_NAME_SIZE: usize = MAX_PATH as usize + 1;
+
+/// Returns a list of zero-terminated wide strings containing volume GUID paths.
+/// Volume GUID paths have the form `\\?\{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}\`.
+///
+/// Rather confusingly, the Win32 API _also_ calls these "volume names".
+pub(crate) fn get_volume_guid_paths() -> Vec<Vec<u16>> {
+    let mut volume_names = Vec::new();
+    unsafe {
+        let mut buf = Box::new([0u16; VOLUME_NAME_SIZE]);
+        let handle = match FindFirstVolumeW(&mut buf[..]) {
+            Ok(handle) => handle,
+            Err(_) => {
+                sysinfo_debug!("Error: FindFirstVolumeW() = {:?}", GetLastError());
+                return Vec::new();
+            }
+        };
+        volume_names.push(from_zero_terminated(&buf[..]));
+        loop {
+            match FindNextVolumeW(handle, &mut buf[..]) {
+                Ok(_) => (),
+                Err(_) => {
+                    let find_next_err = GetLastError().expect_err(
+                        "GetLastError should return an error after FindNextVolumeW returned zero.",
+                    );
+                    if find_next_err.code() != ERROR_NO_MORE_FILES.to_hresult() {
+                        sysinfo_debug!("Error: FindNextVolumeW = {}", find_next_err);
+                    }
+                    break;
+                }
+            }
+            volume_names.push(from_zero_terminated(&buf[..]));
+        }
+        if FindVolumeClose(handle) != Ok(()) {
+            sysinfo_debug!("Error: FindVolumeClose = {:?}", GetLastError());
+        };
+    }
+    volume_names
+}
+
+/// Given a volume GUID path (`\\?\{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}\`), returns all
+/// volume paths (drive letters and mount paths) associated with it
+/// as zero terminated wide strings.
+///
+/// # Safety
+/// `volume_name` must contain a zero-terminated wide string.
+pub(crate) unsafe fn get_volume_path_names_for_volume_name(
+    volume_guid_path: &[u16],
+) -> Vec<Vec<u16>> {
+    let volume_guid_path = PCWSTR::from_raw(volume_guid_path.as_ptr());
+
+    // Initial buffer size is just a guess. There is no direct connection between MAX_PATH
+    // the output of GetVolumePathNamesForVolumeNameW.
+    let mut path_names_buf = vec![0u16; MAX_PATH as usize];
+    let mut path_names_output_size = 0u32;
+    for _ in 0..10 {
+        match GetVolumePathNamesForVolumeNameW(
+            volume_guid_path,
+            Some(path_names_buf.as_mut_slice()),
+            &mut path_names_output_size)
+            .map_err(|_| GetLastError()
+                .expect_err("GetLastError should return an error after GetVolumePathNamesForVolumeNameW returned zero.")
+                .code()) {
```

Comment:
The code is quite difficult to read as is. What do you think about (you will need to run `cargo fmt` on it ^^'):

```suggestion
    let volume_path_names = GetVolumePathNamesForVolumeNameW(
        volume_guid_path,
        Some(path_names_buf.as_mut_slice()),
        &mut path_names_output_size);
    let code = volume_path_names
        .map_err(|_| match GetLastError() {
            Ok(_) => {
                sysinfo_debug!("GetLastError should return an error after GetVolumePathNamesForVolumeNameW returned zero.");
                // We return an error in any case that is not `ERROR_MORE_DATA` to stop the 
                ERROR_UNRECOGNIZED_VOLUME
            }
            Err(e) => e.code(),
        });
        match code {
```

?

I also changed how `GetLastError` error code is handled to ensure it can never panic.

Reply:
Yep, your version is more readable. I was leaning heavily on `let-else` in an earlier version of the code. It makes these manual error condition checks _much_ easier to read 😉. But that would require MSRV 1.65 🥲

As for the `GetLastError` I agree that we should probably not panic. At first, I was really confused by `GetLastError` returning a `Result`, but [according to Microsoft](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror#return-value), there _are_ indeed functions that have you _always_ call `GetLastError()`, which then _sometimes_ returns `0` to indicate success ¯\_(ツ)_/¯

Reply:
> But that would require MSRV 1.65 🥲

I didn't know it was stable. I don't mind updating the MSRV to 1.65 but let's do it in a follow-up PR to prevent changing to many thing in this one. ;)

> As for the `GetLastError` I agree that we should probably not panic. At first, I was really confused by `GetLastError` returning a `Result`, but [according to Microsoft](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror#return-value), there _are_ indeed functions that have you _always_ call `GetLastError()`, which then _sometimes_ returns `0` to indicate success ¯\\\_(ツ)\_/¯

Oh well, windows API is weird, nothing new. ^^'